### DISCUSSION TITLE
Accessing injected dependency from pinia plugin in setup function

### DISCUSSION THREAD

**Discussion by kftang** [vuejs/pinia#1611] [Help and Questions] [answered]
How do I access an injected dependency? I'm using pinia.use() to provide a dependency (wasm module), and I need to use the dependency in my store.

How do I access the dependency if my stores are setup stores? `this` isn't defined since I'm using a setup function

EX:

```
pinia.use(({ store }) => {
    store.lib = markRaw(libModule);
});
```

How can I access this inside:
```
defineStore('store', () => {
// how to get lib here
});
```

**Comment by kftang** [comment]
I ended up getting around this by just making an empty store and accessing the lib from it, but I feel like there is a better way still.

plugin:
```
    pinia.use(() => ({ lib: libModule }));
```

lib store:
```
export const useLibStore = defineStore('lib', {});
```

and in the store using lib:
```
defineStore('store', () => {
    const { lib } = useLibStore();
});
```

**Reply by bnachtweh** [reply]
Yeah this works (for now), but I share your opinion on the matter. I would expect that I can (optionally) expose the stores "context" to my module, just like with the [Vue Component setup](https://vuejs.org/api/composition-api-setup.html#setup-context). In Pinia modules, it could look something like:

```
const useMyStore = defineStore('myStore', (context) => {
    context.lib.doSomething();
})
```

But Pinia Store Setup does not accept function parameters at the moment. Would be great though.

**Reply by nterray** [reply]
I agree, accessing injected content inside the setup function would be very helpful.

**Reply by Shiva127** [reply]
Wow, 3+ years later and still no easy solution. 🤦

**Reply by Ragnar-Oock** [reply]
I found myself with the need to inject stuff in the setup of store too, so I opened #3115 to make it possible.

**Reply by posva** [reply, marked as answer]
Your solution is great. For setup stores, a simple app.provide and then inject within the setup store is the way!

**Comment by aymenhmaidiwastaken** [comment]
In a setup store, you can't access plugin-injected properties via `this` because there's no `this`. Instead, access the store instance explicitly.

**Option 1 — Import the raw Pinia instance and read from it:**

```ts
import { getActivePinia } from 'pinia';

export const useMyStore = defineStore('myStore', () => {
  function doSomething() {
    const pinia = getActivePinia();
    const lib = pinia._s.get('myStore')?.lib; // not ideal
  }

  return { doSomething };
});
```

**Option 2 (Recommended) — Use `provide/inject` at the Vue app level instead of a Pinia plugin:**

```ts
// main.ts
import { provide } from 'vue';

app.provide('wasmLib', libModule);

// In your store — use a composable
import { inject } from 'vue';

export const useMyStore = defineStore('myStore', () => {
  const lib = inject('wasmLib');

  function processData(data: Uint8Array) {
    return lib.process(data);
  }

  return { processData };
});
```

This works because setup stores run in a Vue component context where `inject` is available.

**Option 3 — Simple module-level export:**

If the WASM module is a singleton that doesn't need reactivity:

```ts
// lib.ts
export const lib = await initWasmModule();

// store.ts
import { lib } from './lib';

export const useMyStore = defineStore('myStore', () => {
  // use lib directly
});
```

Option 2 is cleanest if you need it scoped per-app. Option 3 is simplest if it's a global singleton.
